In [1]:
# Importing Libraries - pandas
import pandas as pd

In [2]:
# Reading the customer dataset
cust_df = pd.read_csv("C:\\Users\\daivi\\Desktop\\AML 1114 - Data science and machine learning\\Code and Datasets\\WEEK 3\\2_customers.csv")

In [3]:
# Reading the transactions dataset
trans_df = pd.read_csv("C:\\Users\\daivi\\Desktop\\AML 1114 - Data science and machine learning\\Code and Datasets\\WEEK 3\\3_transactions.csv")

In [4]:
# Printing the top 5 records of customer dataset
cust_df.head()

id       cust_name  num_products province  len_relationship  \
0  1001  Johnny Awesome             3       ON                 1   
1  1002      Bob Marley             4       ON                 1   
2  1004    Taylor Swift             2       AB                 2   
3  1101      James Dean             3       BC                 3   
4  1006   Stephen Smith             5       ON                 4   

      cust_category  
0            wealth  
1  personal_banking  
2               hnw  
3            wealth  
4               hnw

In [5]:
# Rename column id to cust_id
cust_df.rename(columns = {'id' : 'cust_id'}, inplace = True)

In [6]:
cust_df.head()

cust_id       cust_name  num_products province  len_relationship  \
0     1001  Johnny Awesome             3       ON                 1   
1     1002      Bob Marley             4       ON                 1   
2     1004    Taylor Swift             2       AB                 2   
3     1101      James Dean             3       BC                 3   
4     1006   Stephen Smith             5       ON                 4   

      cust_category  
0            wealth  
1  personal_banking  
2               hnw  
3            wealth  
4               hnw

In [7]:
# Lets set the id as the index, otherwise we will face issues while merging
cust_df = cust_df.set_index('cust_id')

In [8]:
cust_df.head()

cust_name  num_products province  len_relationship  \
cust_id                                                            
1001     Johnny Awesome             3       ON                 1   
1002         Bob Marley             4       ON                 1   
1004       Taylor Swift             2       AB                 2   
1101         James Dean             3       BC                 3   
1006      Stephen Smith             5       ON                 4   

            cust_category  
cust_id                    
1001               wealth  
1002     personal_banking  
1004                  hnw  
1101               wealth  
1006                  hnw

In [9]:
# Printing the top 5 records of transaction dataset
trans_df.head()

id  customer_id txn_type  txn_total
0   1         1001    Debit     165.78
1   2         1001   Credit      42.10
2   4         1001   Credit     103.03
3   5         1001   Credit      56.60
4   6         1002    Debit     214.34

In [10]:
# Merging the datasets
merged_df = pd.merge(
    left = trans_df,
    right = cust_df,
    left_on = 'customer_id',
    right_index = True
)


In [11]:
# Printing the top 5 records of merged dataframe
merged_df.head()

id  customer_id txn_type  txn_total       cust_name  num_products province  \
0   1         1001    Debit     165.78  Johnny Awesome             3       ON   
1   2         1001   Credit      42.10  Johnny Awesome             3       ON   
2   4         1001   Credit     103.03  Johnny Awesome             3       ON   
3   5         1001   Credit      56.60  Johnny Awesome             3       ON   
4   6         1002    Debit     214.34      Bob Marley             4       ON   

   len_relationship     cust_category  
0                 1            wealth  
1                 1            wealth  
2                 1            wealth  
3                 1            wealth  
4                 1  personal_banking

### Q1. Write a function to display a Series object with only the names of Wealth customers.

In [12]:
cust_df[cust_df['cust_category'] == 'wealth'] ['cust_name']

cust_id
1001    Johnny Awesome
1101        James Dean
1285        Maria Alva
1585      Steven Prime
Name: cust_name, dtype: object

<b>The wealthy customers are:<br>
1. Johnny Awesome <br>
2. Maria Alva <br>
3. Steven Prime <br>
4. James Dean

### Q2. Write a function to identify all PersonalBanking customers with more than 3 products.

In [13]:
cust_df[(cust_df['cust_category'] == 'personal_banking') & (cust_df['num_products'] > 3)] [['cust_name', 'num_products', 'cust_category']]

cust_name  num_products     cust_category
cust_id                                              
1002       Bob Marley             4  personal_banking
1354     Jessica Fast             4  personal_banking

The customers who are Personal Banking Customers with more than 3 products are - Bob Marley and Jessica Fast

### Q3. What is the average spend of customers in Alberta?

In [14]:
# Printing the top 5 records of merged dataframe
merged_df.head()

id  customer_id txn_type  txn_total       cust_name  num_products province  \
0   1         1001    Debit     165.78  Johnny Awesome             3       ON   
1   2         1001   Credit      42.10  Johnny Awesome             3       ON   
2   4         1001   Credit     103.03  Johnny Awesome             3       ON   
3   5         1001   Credit      56.60  Johnny Awesome             3       ON   
4   6         1002    Debit     214.34      Bob Marley             4       ON   

   len_relationship     cust_category  
0                 1            wealth  
1                 1            wealth  
2                 1            wealth  
3                 1            wealth  
4                 1  personal_banking

In [15]:
# Calculating the mean directly
merged_df['txn_total'].mean()

207.29676470588237

In [16]:
# Average spend of all customers from Alberta
print(f"The average spend of all customers from Alberta is:  { round(merged_df[merged_df['province'] == 'AB'] ['txn_total'].mean(), 2)}")

The average spend of all customers from Alberta is:  208.98


### Q4. How many personal banking customers are there in each province?

In [24]:
print(f"The number of personal banking customers in each province: \n{merged_df[(merged_df['cust_category'] == 'personal_banking')].groupby('province').size()}")

The number of personal banking customers in each province: 
province
AB    3
BC    6
ON    6
dtype: int64


### Q5. Who are the top 3 customers with the highest spend?

In [37]:
# Lets first take the total_spend of each customer
total_spend = merged_df.groupby('cust_name') ['txn_total'].sum()
print(total_spend)

cust_name
Bob Marley            351.40
Delilah Avery         700.43
Jessica Fast          909.87
Johnny Awesome        367.51
Johnson Cory          506.26
Maria Alva            811.40
Mariah Anita Smith    801.69
Robert Optimus        599.19
Stephanie Element     682.40
Stephen Smith         225.89
Steven Prime          858.55
Taylor Swift          233.50
Name: txn_total, dtype: float64


In [41]:
# Now lets sort it and fetch the bottom 3 values as sort_values function arranges the values in ascending order
total_spend.sort_values().tail(3)

cust_name
Maria Alva      811.40
Steven Prime    858.55
Jessica Fast    909.87
Name: txn_total, dtype: float64

**The top 3 customers with the highest spend:** <br>
Maria Alva    -  811.40 <br>
Steven Prime  -  858.55 <br>
Jessica Fast  -  909.87

### Q6. Which customer has the highest single transaction?

In [42]:
merged_df.groupby('cust_name') ['txn_total'].max()

cust_name
Bob Marley            214.34
Delilah Avery         449.39
Jessica Fast          438.29
Johnny Awesome        165.78
Johnson Cory          255.21
Maria Alva            393.91
Mariah Anita Smith    403.62
Robert Optimus        242.73
Stephanie Element     328.72
Stephen Smith         225.89
Steven Prime          320.40
Taylor Swift          227.58
Name: txn_total, dtype: float64

In [44]:
merged_df[merged_df['txn_total'] == merged_df['txn_total'].max()] [['cust_name', 'txn_total']]

cust_name  txn_total
18  Delilah Avery     449.39

Delilah Avery is the highest single transaction of 449.39

### Q7. What is the average spend by Province and Customer Category?

In [48]:
merged_df.groupby(['province', 'cust_category'])['txn_total'].mean()

province  cust_category   
AB        hnw                 147.952000
          personal_banking    233.476667
          wealth              286.183333
BC        personal_banking    285.260000
MB        hnw                 199.730000
ON        hnw                 225.890000
          personal_banking    172.300000
          wealth              168.415714
Name: txn_total, dtype: float64

### Q8. What is the average spend by transaction type (e.g. credit, debit, cash)?

In [49]:
merged_df.head()

id  customer_id txn_type  txn_total       cust_name  num_products province  \
0   1         1001    Debit     165.78  Johnny Awesome             3       ON   
1   2         1001   Credit      42.10  Johnny Awesome             3       ON   
2   4         1001   Credit     103.03  Johnny Awesome             3       ON   
3   5         1001   Credit      56.60  Johnny Awesome             3       ON   
4   6         1002    Debit     214.34      Bob Marley             4       ON   

   len_relationship     cust_category  
0                 1            wealth  
1                 1            wealth  
2                 1            wealth  
3                 1            wealth  
4                 1  personal_banking

In [50]:
merged_df.groupby('txn_type') ['txn_total'].mean()

txn_type
Cash      182.262727
Credit    206.926667
Debit     227.204286
Name: txn_total, dtype: float64

### Q9. Which transaction type is the most popular in each province?

In [55]:
merged_df.groupby('province')['txn_type'].max()

province
AB     Debit
BC     Debit
MB    Credit
ON     Debit
Name: txn_type, dtype: object